mm2src/mm2_main/src/lp_ordermatch.rs

```diff
@@ -2624,6 +2731,268 @@ fn collect_orderbook_metrics(ctx: &MmArc, orderbook: &Orderbook) {
+                        },
+                        Duration::from_secs(TRIE_STATE_HISTORY_TIMEOUT),
+                    );
+
+                    pubkey_state
+                        .order_pairs_trie_state_history
+                        .get_mut_unchecked(&alb_pair_owned)
+                        .expect("must exist")
+                },
+            };
+
+            history.insert_new_diff(
+                prev_root,
+                TrieDiff {
+                    delta: vec![(uuid, Some(order))],
+                    next_root: *pair_root,
+                },
+            );
+        }
+    }
+
+    fn apply_remove(&mut self, pubkey: &str, alb_pair: &str, uuid: Uuid) {
+        let pubkey_state = pubkey_state_mut(&mut self.pubkeys_state, pubkey);
+        let pair_state = order_pair_root_mut(&mut pubkey_state.trie_roots, alb_pair);
+        let old_state = *pair_state;
+
+        let to_remove = &(uuid, alb_pair.to_owned());
+        pubkey_state.orders_uuids.remove(to_remove);
+
+        if old_state == H64::default() || old_state == hashed_null_node::<Layout>() {
+            return;
+        }
+
+        *pair_state = match delta_trie_root::<Layout, _, _, _, _, _>(
+            &mut self.memory_db,
+            *pair_state,
+            vec![(*uuid.as_bytes(), None::<Vec<u8>>)],
+        ) {
+            Ok(root) => root,
+            Err(_) => {
+                error!("Failed to get existing trie with root {:?}", pair_state);
+                return;
+            },
+        };
+
+        let alb_pair_owned = alb_pair.to_owned();
+
+        let _ = pubkey_state
+            .order_pairs_trie_state_history
+            .update_expiration_status(alb_pair_owned.clone(), Duration::from_secs(TRIE_STATE_HISTORY_TIMEOUT));
+
+        if let Some(history) = pubkey_state
+            .order_pairs_trie_state_history
+            .get_mut_unchecked(&alb_pair_owned)
+        {
+            history.insert_new_diff(
+                old_state,
+                TrieDiff {
+                    delta: vec![(uuid, None)],
+                    next_root: *pair_state,
+                },
+            );
+        }
+    }
+
+    fn apply_clear_pair(&mut self, pubkey: &str, alb_pair: &str) {
+        if let Some(pubkey_state) = self.pubkeys_state.get_mut(pubkey) {
+            pubkey_state.order_pairs_trie_state_history.remove(&alb_pair.to_owned());
+
+            pubkey_state.orders_uuids.retain(|(_uuid, pair)| pair != alb_pair);
+
+            pubkey_state.trie_roots.remove(alb_pair);
+        }
+    }
+
+    /// Build a SyncPubkeyOrderbookState request if keep-alive indicates our local trie roots are stale.
+    /// This avoids touching Orderbook (and its lock). Topic subscription is provided via `is_subscribed`.
+    fn prepare_sync_request_for_keep_alive(
+        &mut self,
+        from_pubkey: &str,
+        message: new_protocol::PubkeyKeepAlive,
+        i_am_relay: bool,
+        mut is_subscribed: impl FnMut(&str) -> bool,
```

Comment:
Why is this needed to be mutable?

Reply:
No, fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2661/commits/f1f4f4824add28f8aa850fba44c9a0706c989a48